src/Build/BackEnd/Components/Logging/ILoggingService.cs

```diff
@@ -650,6 +655,16 @@ bool HaveLoggedBuildFinishedEvent
             get;
             set;
         }
+
+        /// <summary>
+        /// Has the sink logged the BuildCanceledEvent. This is important to know because we only want to log the build canceled event once
+        /// </summary>
+        bool HaveLoggedBuildCanceledEvent
+        {
+            get;
+            set;
+        }
+
```

Comment:
Is this state realy needed in the logging infra (LoggingService and EventSourceSink)?
It feels that the logging infra responsibility should be to only transfer the events that have been logged. If there needs to be any resilience against multiple occurence of events - than it's likely a responsibility of a concrete logger.

Let's imagine we have a bug occuring when cancel request is issued multiple times - it'd be good to see in the binlog what exactly happened - that there were multiple cancel requests.

Appart from that - it would prevent bloating responsibilities of already bloated LoggingService (I know this wouldn't be a first state that's special case held in LoggingService - but let's not stick to that)

Reply:
I don't have strong opinion here. I would be more in favor of centralized implementation than copy similar code across multiple loggers, but it could be in form of some filter. Not familiar with the logging infra end-to-end, so I think we should discuss and cover this in different issue.
I would like to (for example) see use of observer pattern instead of C# events (which would handle event duplicates easily), but the goal of this PR is to fix cancellation UX. Don't get me wrong, there are places that would deserve some refactoring and simplification, but I would really cover it separately.

cc: @rainersigwald

Reply:
We decided that cancellation event doesn't need special treatment.